hibernate-core/src/main/java/org/hibernate/query/internal/AbstractProducedQuery.java

```diff
@@ -749,57 +749,86 @@ public boolean isBound(Parameter<?> parameter) {
+				boundCheckingLogMessage
+		);
 	}
 
 	@Override
 	public Object getParameterValue(String name) {
 		getProducer().checkOpen( false );
 
-		final QueryParameterBinding binding;
-		try {
-			binding = getQueryParameterBindings().getBinding( name );
-		}
-		catch (QueryParameterException e) {
-			throw new IllegalArgumentException( "Could not resolve parameter by name - " + name, e );
-		}
+		final String notBoundErrorMessage = "Parameter value not yet bound : " + name;
+		final String couldNotResolveErrorMessage = "Could not resolve parameter by name - " + name;
+		final String boundCheckingLogMessage = "Checking whether named parameter [ " + name + "] is bound : %s";
 
-		LOGGER.debugf( "Checking whether named parameter [%s] is bound : %s", name, binding.isBound() );
-		if ( !binding.isBound() ) {
-			throw new IllegalStateException( "Parameter value not yet bound : " + name );
-		}
-		return binding.getBindValue();
+		final QueryParameter<Object> queryParameter = getParameterMetadata().getQueryParameter( name );
+		return getParameterValue(
+				queryParameter,
+				notBoundErrorMessage,
+				couldNotResolveErrorMessage,
+				boundCheckingLogMessage
+		);
 	}
 
 	@Override
 	public Object getParameterValue(int position) {
 		getProducer().checkOpen( false );
 
-		final QueryParameterBinding binding;
+		final String notBoundErrorMessage = "Parameter value not yet bound : " + position;
+		final String couldNotResolveErrorMessage = "Could not resolve parameter by position - " + position;
+		final String boundCheckingLogMessage = "Checking whether positional parameter [ " + position + "] is bound : %s";
+
+		final QueryParameter<Object> queryParameter = getParameterMetadata().getQueryParameter( position );
+		return getParameterValue(
+				queryParameter,
+				notBoundErrorMessage,
+				couldNotResolveErrorMessage,
+				boundCheckingLogMessage
+		);
+	}
+
+	private Object getParameterValue(
+			QueryParameter queryParameter,
+			String notBoundErrorMessage,
+			String couldNotResolveErrorMessage,
+			String boundCheckingLogMessage) {
 		try {
-			binding = getQueryParameterBindings().getBinding( position );
+			final QueryParameterBindings parameterBindings = getQueryParameterBindings();
+
+			if ( queryParameter == null ) {
+				throw new IllegalArgumentException( couldNotResolveErrorMessage );
+			}
+			if ( parameterBindings.isAParameterList( queryParameter ) ) {
+				final QueryParameterListBinding<Object> queryParameterListBinding = parameterBindings
+						.getQueryParameterListBinding( queryParameter );
+				final Collection<Object> bindValues = queryParameterListBinding.getBindValues();
+				if ( bindValues == null ) {
+					throw new IllegalStateException( notBoundErrorMessage );
+				}
+				return bindValues;
+			}
+
+			final QueryParameterBinding<Object> binding = parameterBindings.getBinding( queryParameter );
+			final boolean bound = binding.isBound();
+			LOGGER.debugf( boundCheckingLogMessage, bound );
```

Comment:
We should pay attention to the performance cost for the above statement as well. The first parameter building cost will be paid even when we disable debug level logging!

Reply:
Thanks @NathanQingyangXu ,I'll try to fix this.